File: python/1376_Time_Needed_to_Inform_All_Employees.py
class Solution:
    def numOfMinutes(self, n: int, head: int, manager: List[int], informTime: List[int]) -> int:
        g = [[] for _ in range(n)]
        for i in range(n):
            if i == head:
                continue
            g[manager[i]].append(i)
        
        q = deque([(head, 0)])

        ans = 0

        while q:
            u, t = q.popleft()
            if not g[u]:
                ans = max(ans, t)
                continue    
            for v in g[u]:
                q.append((v, t + informTime[u]))
        
        return ans
        

File: cpp/1376_Time_Needed_to_Inform_All_Employees.cpp
#include <vector>
#include <queue>
#include <cstring>

using namespace std;

class Solution {
public:
    int numOfMinutes(int n, int head, vector<int>& manager, vector<int>& informTime) {
        vector<vector<int>> g(n);
        for (int i = 0; i < n; i++) {
            if (i == head) {
                continue;
            }
            g[manager[i]].push_back(i);
        }
        vector<int> t(n, 0);
        queue<int> q;
        q.push(head);
        int ans = 0;
        while (!q.empty()) {
            int u = q.front();
            q.pop();
            if (g[u].empty()) {
                ans = max(ans, t[u]);
                continue;
            }
            for (int v : g[u]) {
                q.push(v);
                t[v] = t[u] + informTime[u];
            }
        }
        return ans;
    }
};